Feature/boolti 405 검색탭 UI 작업

mangbaam opened the pull request:
## Issue
- close #405 

## 작업 내용

- 검색 탭, 최근 검색 화면, 검색어 추천 화면, 검색 결과 화면 등 UI 작업
- MVI 스타일로 한번 해보았어요 (제대로 한 건 아님)
- Repository 에 목업 데이터로 일단 처리

### 후속 작업

- Show, User.Others 등 도메인 모델을 UI 모델로 매핑
- 실제 API 연동

https://github.com/user-attachments/assets/a2420437-5daa-47fc-950e-ad5d36100c93

mangbaam committed:
Boolti-405 탐색 관련 화면 추가 및 화면 이동 구현

mangbaam committed:
Boolti-405 BooltiDatabase 추가

mangbaam committed:
Boolti-405 탐색 Repository 추가

mangbaam committed:
Boolti-405 최근 검색 화면 추가

mangbaam committed:
Boolti-405 BtSearchBar 클리어버튼 추가

mangbaam committed:
Boolti-405 검색에 trim 추가

mangbaam committed:
Boolti-405 최근검색어 이탈할 때 검색어 제거하기

mangbaam committed:
Boolti-405 검색탭 새로 등록된 공연까지 UI 작업

mangbaam committed:
Boolti-405 검색탭 요즘 뜨는 검색어 UI 작업

mangbaam committed:
Boolti-405 검색어 워싱 및 하이라이트 UI 작업

mangbaam committed:
Boolti-405 검색 결과 없을 때 UI

mangbaam committed:
Boolti-405 검색 결과화면에서 뒤로가기 하면 항상 최근 검색화면 보이도록 수정

mangbaam committed:
Boolti-405 검색 결과화면 인텐트 추가

mangbaam committed:
Boolti-405 검색 결과화면 탭 추가

mangbaam committed:
Boolti-405 검색 결과화면 탭 결과 없음 UI 추가

mangbaam committed:
Boolti-405 검색 결과화면 공연탭 추가

mangbaam committed:
Boolti-405 검색 결과화면 아티스트탭 추가

mangbaam committed:
Boolti-405 검색 결과화면 전체탭 추가

mangbaam committed:
Boolti-405 검색 결과화면 로딩 아닐 때만 컨텐츠 노출

mangbaam committed:
Boolti-405 검색탭 Repository 연결

mangbaam committed:
Boolti-405 검색탭 검색어 추천 Repository 연결

mangbaam committed:
Boolti-405 검색 Repository 연결

HamBP commented on presentation/src/main/java/com/nexters/boolti/presentation/extension/Char.kt:
```diff
@@ -0,0 +1,31 @@
+ *
+ * @return 한글 자음이면 true, 아니면 false
+ */
+fun Char.is자음(): Boolean = this in '\u3131'..'\u314E'
```
오 한글 함수다

[그냥 궁금] 정규표현식 대신 유니코드를 쓴 이유가 있을까?

mangbaam replied:
Char 하나를 비교하는거라 메모리 할당도 필요없어서 유니코드 범위 파악이 훨씬 효율적이라서!
이정도 로직에 정규식 사용하면 불필요하게 메모리 할당하고 패턴매칭하고 암튼 오버헤드가 커

mangbaam committed:
[Boolti-410] 라이브러리 최신화

- kotlin: 2.2.20
- android: 8.13.0
- ksp: 2.2.20-2.0.4
- hilt: 2.57.2
- room: 2.8.2
- gradle: 8.14.3

mangbaam committed:
[Boolti-410] 라이브러리 최신화

- composeBom: 2025.10.00
- material: 1.13.0
- material3Android: 1.4.0
- materialIcons: 1.0.0 (new)

PullToRefresh 변경

mangbaam committed:
[Boolti-410] 라이브러리 최신화

- targetSdk: 36
- compileSdk: 36
- activity-ktx: 1.11.0
- lifecycle: 2.9.4
- appcompat: 1.7.1
- core-ktx: 1.17.0

mangbaam committed:
[Boolti-410] 라이브러리 최신화

- constraintlayout: 2.2.1
- constraintlayoutCompose: 1.1.1
- navigationCompose: 2.9.5
- hiltNavigationCompose: 1.3.0

mangbaam committed:
[Boolti-410] 라이브러리 최신화

- datastore: 1.1.7
- coroutines: 1.10.2
- kotlinxSerializationJson: 1.9.0
- googleServices: 4.4.4
- immutable: 0.4.0

hiltViewModel 의존성 변경

mangbaam committed:
[Boolti-410] 라이브러리 최신화

- coil: 2.7.0
- lottie: 6.6.10

mangbaam committed:
[Boolti-410] 라이브러리 최신화

- retrofit: 3.0.0
- logging-interceptor: 5.2.1

mangbaam committed:
[Boolti-410] 라이브러리 최신화

- zoomable: 2.8.2
- kakao: 2.22.0

mangbaam committed:
[Boolti-410] 라이브러리 최신화

- junit: 1.3.0
- espresso: 3.7.0
- kotest: 6.0.4
- mockk: 1.14.6

mangbaam committed:
[Boolti-410] 라이브러리 최신화

- firebase-bom: 34.4.0
- crashlytics: 3.0.6

mangbaam committed:
[Boolti-410] 티켓탭 바로 안 나오는 버그 수정

mangbaam committed:
Merge pull request #411 from Nexters/feature/Boolti-410

Feature/boolti 410 라이브러리 최신화

HamBP commented on data/src/main/java/com/nexters/boolti/data/db/dao/SearchHistoryDao.kt:
```diff
@@ -0,0 +1,23 @@
+
+@Dao
+interface SearchHistoryDao {
+    @Insert(onConflict = OnConflictStrategy.REPLACE)
```
👍

mangbaam replied:
헤헤

HamBP commented on presentation/src/main/java/com/nexters/boolti/presentation/screen/Main.kt:
```diff
@@ -148,6 +154,31 @@ fun MainNavigation(
+            },
+            navigateUp = {
+                navController.popBackStack()
+                navController.navigate(SearchRoute.RecentSearch) {
```
navController를 CompositionLocal 로 넘겨서 각 스크린에서 사용하는 것으로 변경되었었는데, 다시 이벤트 드릴링 방식으로 바꾼 건가??

mangbaam replied:
아 그랬나? ㅋㅋ 화면이동 오랜만에 짜서 레퍼런스를 잘못 한 듯?

mangbaam replied:
[7a8820a](https://github.com/Nexters/Boolti/pull/409/commits/7a8820adcbc250be57164b78db795ca05bf77138)

HamBP commented on presentation/src/main/java/com/nexters/boolti/presentation/screen/search/SearchUiModel.kt:
```diff
@@ -0,0 +1,25 @@
+    val showClearHistoriesDialog: Boolean,
+) {
+    companion object {
+        val Default = SearchUiModel(
```
👍

mangbaam replied:
헤헤

HamBP commented:
혹시 MVI 만들 때 봤던 아티클이나 리포지토리 있으면 같이 확인해볼게!

내가 MVI 연습했을 때랑은 조금 다른 부분이 있는데, 아마 그 부분이 

> (제대로 한 건 아님)

에 해당하는 거 같긴 한데, 참고 레퍼런스가 있으면 공유해 주면 좋을 듯!

예를 들면, 엄격한 MVI 코드에선 (아닐 수도 있음) onIntent 메서드에서만 상태를 변경할 수 있도록 제한을 두기도 하는데, 예시 코드는 아래와 같아.

<img width="793" height="838" alt="image" src="https://github.com/user-attachments/assets/6b1de6e1-0dac-4c45-bb10-2506686a9714" />
(출처: https://charlezz.com/?p=46365)

근데 이 코드의 경우 몇몇 케이스에 대해 단순한 로직도 복잡하게 만드는 경향이 있기도 해. 그래서 명범이 원하는 제약 정도(?)가 어느정도인지 코드의 추구미(?)가 어느쪽인지 얘기해 보면 좋을 듯!

HamBP commented on presentation/src/main/java/com/nexters/boolti/presentation/screen/search/SearchScreen.kt:
```diff
@@ -0,0 +1,411 @@
+            modifier = Modifier
+                .defaultMinSize(minHeight = 48.dp)
+                .fillMaxWidth()
+                .padding(horizontal = marginHorizontal)
```
Divider의 패딩이랑 Row의 padding이 중복으로 들어간 거 같아!

디자인 시안
<img width="652" height="610" alt="image" src="https://github.com/user-attachments/assets/efa992ba-ee71-4476-af69-00b2c3991c3e" />

렌더링
<img width="431" height="330" alt="image" src="https://github.com/user-attachments/assets/c6e2ac8c-0b1d-4ca6-8d29-7708be10c501" />

mangbaam replied:
[c84c299](https://github.com/Nexters/Boolti/pull/409/commits/c84c299f179ab8800d78f81f41da9053e2ae941b)

HamBP commented on presentation/src/main/java/com/nexters/boolti/presentation/screen/home/HomeScreen.kt:
```diff
@@ -128,9 +128,15 @@ fun HomeScreen(
             )
 
+            searchScreen(
+                modifier = Modifier.padding(bottom = innerPadding.calculateBottomPadding()),
```
이쪽인지는 모르겠는데, 어딘가에 들어간 패딩 때문에 검색어가 잘리는 거 같아!

<img width="574" height="339" alt="image" src="https://github.com/user-attachments/assets/ee0b1eb5-db95-405f-bd95-fa7c56facdf0" />

mangbaam replied:
오 내비게이션을 제스쳐 말고 버튼으로 하니까 아래 패딩이 생기네
굿굿

mangbaam replied:
[2b5c3fe](https://github.com/Nexters/Boolti/pull/409/commits/2b5c3fe174445b0e97c7b26903336ff00b183233)

HamBP commented on presentation/src/main/java/com/nexters/boolti/presentation/screen/search/recent/RecentSearchUiState.kt:
```diff
@@ -0,0 +1,48 @@
+) {
+    val showClearButton: Boolean = recentSearchKeywords.size >= 2
+
+    // 검색 효율을 높이기 위해 마지막 한 글자가 한글자모인 경우 마지막 글자를 제외하고 검색
```
👍

mangbaam replied:
남준사마가 행복해하면 좋겠다

mangbaam commented:
(일단 따로 참고한 MVI 자료는 없긴 해 + 그리고 아래는 순전히 나의 생각)

MVI 를 비롯한 코드 패턴이나 아키텍처에 대한 나의 쭈꾸미🐙는 그 목적성 + 실효성인 거 같아.

예전에 리뷰로 올려준 포스트도 본 적 있고 여러 글들을 봤지만, 저 방식의 목적은 확실하지만 실효성이 있는지는 잘 모르겠어.

올려준 내용에 의하면 Intent(Event) 를 통해서만 UI 상태가 바뀌어야 하는 제약이 걸리게 되는 것 같은데, 실제론 UI 에서 넘어오는 Intent 뿐만 아니라 외부 상태의 변화, 이벤트 등에 의해서도 UI 는 충분히 바뀔 수 있고, 동일하게 UI 가 바뀌더라도 찍어야 하는 로그의 내용이 달라야 한다거나 한다면 그러면 Reducer 에서 동일한 내용으로 UiState 를 업데이트 하더라도 그때마다 Intent 를 분리해야 하는건가? 뷰모델 내에서나 혹은 앞서 말한 것 처럼 UI 가 아닌 외부의 변화로 인해 변경이 발생해야 하면 ViewModel 내에서도 모두 Intent 로만 소통해야하는 건가? 그러면 UI 상태를 업데이트 하는 코드를 한 곳으로 모은다고 가독성이 좋아지는 건가? 응집도가 좋다고 할 수 있는건가? UI 상태를 업데이트 하는 것 외의 다른 코드는 파악하기 쉬워지는 걸까?

엄격한(?) MVI 패턴을 구현하면 이런 의문점들이 생길 수밖에 없는 것 같아. 그래서 이번 작업의 선은 "UI 에 직접 뷰모델의 함수를 노출하지 말고 Intent 만 전달하자. 뷰모델에서 Intent 를 받으면 거기에 대응한 처리를 해주자"가 목적이었고, Intent 자체도 주로 유저의 액션에 의해 발생하는 걸로 생각하고 있었어. (이 기준은 상황에 따라 달라질 수도 있을 듯)

암튼 이게 MVI 가 맞다 틀리다를 논하기보단 준영이 말한대로 의견을 나눠보고 서로 납득 가능한 선에서 싱크하는게 훨씬 의미있는 일이라고 생각해

HamBP commented on presentation/src/main/java/com/nexters/boolti/presentation/screen/search/detail/SearchDetailViewModel.kt:
```diff
@@ -0,0 +1,90 @@
+
+            launch(Dispatchers.IO) {
+                searchHistoryRepository.saveSearchHistory(keyword)
+            }
```
이거 아마 룸에서 디스패처 전환 해줄 거 같습니다! 안 해주더라도 IO 전환하는 건 데이터 레이어 쪽에서 해 주는 게 좀 더 깔끔할 거 같아!

mangbaam replied:
[86948c1](https://github.com/Nexters/Boolti/pull/409/commits/86948c19d9841beb532029d0b0586089f82a4f8c)

HamBP reviewed (approved):
고...고생했어... 명범....

api 나오면 그것만 연결하면 될 듯
그거는 머... 안 봐도 되겠지~

mangbaam committed:
[Boolti-410] 요즘 뜨는 검색어 구분선 길이 수정

mangbaam committed:
[Boolti-410] 탐색 탭 중복 Scaffold 제거

mangbaam committed:
[Boolti-410] 디스패처 제거

mangbaam committed:
[Boolti-410] navController 참조 위치 변경

mangbaam committed:
[Boolti-405] 탐색 홈 API 연결

mangbaam committed:
[Boolti-405] 공연 페이징 API 추가

mangbaam committed:
[Boolti-405] 아티스트 페이징 API 추가

mangbaam committed:
[Boolti-405] 검색 공연, 아티스트 탭 무한스크롤

mangbaam committed:
[Boolti-405] 무한스크롤 컴포저블 추출

mangbaam committed:
[Boolti-405] 홈에 검색창 제거

mangbaam committed:
[Boolti-405] API 스펙 변경 대응

mangbaam committed:
[Boolti-405] 페이징 다음 페이지 없을 때 호출 X

mangbaam committed:
[Boolti-405] 자동완성 키워드 API 연결

mangbaam committed:
Merge pull request #413 from Nexters/feature/Boolti-405-api

Feature/boolti 405 탐색 API 연결

mangbaam committed:
Merge branch 'develop' into feature/Boolti-405

mangbaam committed:
[Boolti-405] 미사용 코드 제거
